test/mcmc/gibbs_new.jl

```diff
@@ -0,0 +1,176 @@
+        vns = DynamicPPL.TestUtils.varnames(model)
+        # Run one sampler on variables starting with `s` and another on variables starting with `m`.
+        vns_s = filter(vns) do vn
+            DynamicPPL.getsym(vn) == :s
+        end
+        vns_m = filter(vns) do vn
+            DynamicPPL.getsym(vn) == :m
+        end
+
+        samplers = [
+            GibbsV2(
+                vns_s => NUTS(),
+                vns_m => NUTS(),
+            ),
+            GibbsV2(
+                vns_s => NUTS(),
+                vns_m => HMC(0.01, 4),
+            )
+        ]
+
+        if !has_dot_assume(model)
+            # Add in some MH samplers, which are not compatible with `.~`.
+            append!(
+                samplers,
+                [
+                    GibbsV2(
+                        vns_s => HMC(0.01, 4),
+                        vns_m => MH(),
+                    ),
+                    GibbsV2(
+                        vns_s => MH(),
+                        vns_m => HMC(0.01, 4),
+                    )
+                ]
+            )
+        end
+
+        @testset "$sampler" for sampler in samplers
+            # Check that taking steps performs as expected.
+            rng = Random.default_rng()
+            transition, state = AbstractMCMC.step(rng, model, DynamicPPL.Sampler(sampler))
+            check_transition_varnames(transition, vns)
+            for _ = 1:5
+                transition, state = AbstractMCMC.step(rng, model, DynamicPPL.Sampler(sampler), state)
+                check_transition_varnames(transition, vns)
+            end
+        end
+    end
+end
+
+@testset "Gibbs using `condition`" begin
+    @testset "demo_assume_dot_observe" begin
+        model = DynamicPPL.TestUtils.demo_assume_dot_observe()
+
+        # Sample!
+        rng = Random.default_rng()
+        vns = [@varname(s), @varname(m)]
+        sampler = GibbsV2(map(Base.Fix2(Pair, MH()), vns)...)
+
+        @testset "step" begin
+            transition, state = AbstractMCMC.step(rng, model, DynamicPPL.Sampler(sampler))
+            check_transition_varnames(transition, vns)
+            for _ = 1:5
+                transition, state = AbstractMCMC.step(rng, model, DynamicPPL.Sampler(sampler), state)
+                check_transition_varnames(transition, vns)
+            end
+        end
+
+        @testset "sample" begin
+            chain = sample(model, sampler, 1000; progress=false)
+            @test size(chain, 1) == 1000
+            display(mean(chain))
+        end
+    end
+
+    @testset "gdemo with CSMC & ESS" begin
+        # `GibbsV2` does not work with SMC samplers, e.g. `CSMC`.
+        # FIXME: Oooor it is (see tests below). Uncertain.
+        Random.seed!(100)
+        alg = GibbsV2(@varname(s) => CSMC(15), @varname(m) => ESS(:m))
+        chain = sample(gdemo(1.5, 2.0), alg, 10_000)
+        @test_broken mean(chain[:s]) ≈ 49 / 24
+        @test_broken mean(chain[:m]) ≈ 7 / 6
```

Comment:
@yebai Any idea why this fails? It does work with `CSMC` in the example below..

Reply:
Main difference I see here is that this example uses CSMC on continuous rather than discrete.

Reply:
(I've also tried without the `ESS(:m)` locally, and it doesn't make a difference)

Reply:
As me and @yebai discussed, this issues is likely caused by the fact that SMC samplers will now also treat the conditioned variables as observations, which is not quite what we want.

But atm I don't quite see how we can work around this with the existing functionality we have in DynamicPPL :confused: 

Basically, we have `fix` and `condition`:
- `condition` makes the variables be treated as observations, which, as seen above, is bad for likelihood-tempered samplers.
- `fix` avoids variables being treated as observations _but_ it also doesn't include the log-prob of that variable. This then causes issues in models with hierarchical dependencies, e.g. `gdemo` where the prior on `m` depends on the value of `s`, and so changing `s` changes the joint also through the prior on `m`.

Effectively, we'd need something similar to `fix` which _also_ computes the log-prob.

Reply:
One thing that confuses me slightly is that, _technically_, even if the SMC sampler sees one of the parameters as an observation, AFAIK, it shouldn't _break_ things, no? It might increase the variance, etc. but it should still converge.

Reply:
And I just tried altering the implementation of `FixedContext` so it also computes the prior and adds it to the logjoint, and then swapped out usage of `condition` in the new Gibbs with `fix`.

Still giving me similarly incorrect answers :confused:

Reply:
I genuinely don't understand what's going on here :confused: 

So, I've addressed the issue of `m` being treated as a observation when running `PG`, but this does not resolve the issue.

I've also tried to inspect the conditional distribution of the two different samplers to see if there's something obvious that pops out ; see below (`k` refers to the number of steps from initial state, and `k = 0` is only the initial state).

There are clearly differences, but I have not been able to identify the cause. Debugging the PG sampler is quite the pain :confused: 

Honestly, it might be worth it to just disallow PG samplers for now and open a corresponding issue which we link in the error message. 

Thoughts @yebai @devmotion ?

![gdemo_m_k=0](https://github.com/TuringLang/Turing.jl/assets/11074788/e70f23b9-40d3-4500-ab2d-40f89ccc96c0)
![gdemo_s_k=0](https://github.com/TuringLang/Turing.jl/assets/11074788/8038c89a-c7c5-43d7-a322-782ff1c84543)
![gdemo_m_k=1](https://github.com/TuringLang/Turing.jl/assets/11074788/ec5a7300-3946-4d70-a8fd-7a9096e263cb)
![gdemo_s_k=1](https://github.com/TuringLang/Turing.jl/assets/11074788/0f123d82-a9bd-4195-8d2a-93c8ed26abb9)
![gdemo_m_k=5](https://github.com/TuringLang/Turing.jl/assets/11074788/586ba964-0098-4621-95c0-1739cb19319a)
![gdemo_s_k=5](https://github.com/TuringLang/Turing.jl/assets/11074788/0fb940b8-58bd-4415-8c45-2943709bb75d)

Reply:
A few things:
1. As mentioned in one of the previous comments, the issue still persists even if I address this issue by using `fix` instead of `condition`, and then altering `fix` to also compute the log-prob (the plots above are generated using this fix). 
2. But AFAIK this should still work, _even without my fix_. It might have higher variance, but it should still converge to the correct thing. " we cannot treat m as data, since that would double-count information in D" I don't understand this. How are we double-counting the data? CSMC should work just fine here; it's just that now we've tacked on one additional distribution in our importance weight, i.e. $p(m \mid s)$, no? The fact that we have one additional resampling step in there shouldn't matter much.

Reply:
Okay, so I think I might know what's up here.

1. SMC samplers uses a separate `VarInfo` for every particle.
2. SMC samplers relies on accumulating log-probs of _everything_ in the tilde-statements.
3. Because of (2), any accumulation of log-probs in anything that is _not_ in the tilde-statements, means that the task-local `varinfo` will _not_ contain this log-prob.

But then this also means that `PG` won't work with models using existing functionality too, e.g. `@addlogprob!` :confused: 

But this does not seem like something we should try to fix in this PR, but requires a more substantial change in DynamicPPL or something.

One potential avenue to address these issues might be to introduce a `addlogprob!!(context, varinfo, logp)`, which can then be overloaded for the scenario of particle samplers, which can then accumulate this in the trask-local `varinfo` instead of the "global" one.

Reply:
Okay, okay, so I've fixed the above issue locally (will make a PR to DPPL), but _still_ this example fails.

_Now_ the issue is with ESS. ESS assumes the provided state (which is just a `VarInfo`) comes from one where only the log-likelihood has been accumulated, while the current impl of `GibbsV2` recomputes the _log-joint_. Hence we the rejection sampling in ESS ends up comparing the likelihood of the proposal to the _joint_ of the previous state.

Will have a look at this later.

Reply:
Aight, so I think we'll need to adopt the same `rerun` functionality as for the current impl of `Gibbs`, since we need to re-evaluate the model in the same fashion as before but without sampling.

Reply:
I haven't followed the discussion in this thread but regarding the last point (ESS): This difference (log-likelihood vs log-joint) was one (the?) main reason for the current design of the re-evaluation in `Gibbs`, i.e., for `gibbs_rerun`: https://github.com/TuringLang/Turing.jl/blob/6649f10c48917a27531214f02777408d2ab82928/src/mcmc/gibbs.jl#L134-L142 We try to avoid re-evaluating `varinfo.logp` when we know that it is not needed but to be safe (and e.g. for `ESS`) the default is to re-evaluate.

Reply:
Aaalrighty! Adopting the rerun mechanisms from the current `Gibbs` seems to have done the trick!

I've also tried a `acclogp!!(context, varinfo, logp)` implementation, and I think this is something we should adopt, as it will immediately address the issue of SMC samplers not currently working with `@acclogprob!!` + there are other scenarios where we might want to hook into this, e.g. `DebugContext` where we can to "record" these hard-coded acclogprobs for debugging and model checking purposes.

Reply:
Btw, @devmotion didn't see your comment because was working on a train without internet :upside_down_face:

Reply:
> One thing that confuses me slightly is that, technically, even if the SMC sampler sees one of the parameters as an observation, AFAIK, it shouldn't break things, no? It might increase the variance, etc. but it should still converge.

To make this absolutely clear, let's look at PG from Andrieu (2010):

![Screenshot_20231121_111042](https://github.com/TuringLang/Turing.jl/assets/11074788/96ef6090-f5e0-4cd9-9028-0ac9c579b28f)

Looking at this, it might seem as if we shouldn't include the "prior" probability of $m$ in `gdemo`, since $p_{\theta}(x_{1:T} \mid y_{1:T})$ does _not_ include $p(\theta)$.

_But_ this is all done under the assumption that the joint nicely factorizes as
$$p(x_{1:T}, \theta \mid y_{1:T}) \propto p_{\theta}(x_{1:T}, y_{1:T}) p(\theta)$$
In this scenario, $p\big(\theta(i)\big)$ will be the same value for all particles in step (b), and so there's no need to include this in the weights.

_But_ in our scenario, the joint does _not_ nicely factorize as above. We have $\theta = m$ and $x_{1:T} = s$, so
$$p(s, m \mid y_{1:T}) = p(s) p(m \mid s) p(y_{1:T} \mid m, s)$$
To get a similar factorization as above, we can of course just define
$$p_m(s, y_{1:T}) = \frac{p(s, m, y_{1:T})}{p(m)} \propto p(s, m, y_{1:T})$$
where we can just drop the contribution of $p(m)$ since it's constant wrt. $s$ that we're sampling.

With this notation, mapping this to the algorithm above is immediate, in which case it's clear that we're targeting the full joint, and so, in the general scenario, we need to include the log-prob contribution of latent variables that are not being sampled by conditional SMC.

And more generally speaking, PG is just a Gibbs sampler with:
1. Using some method (in  Andrieu (2010) they assume it's available in closed-form)
   $$\theta(i) \sim p \big( \theta \mid y_{1:T}, X_{1:T} \propto p(\theta, x_{1:T}, y_{1:T})$$
2. Using CSMC
   $$X_{1:T}(i) \sim p \big( X_{1:T} \mid y_{1:T}, \theta \big) \propto p(\theta, x_{1:T}, y_{1:T})$$

And the reason why we're _not_ including the contribution from the variable we're targeting with CSMC is because we would then have to adjust the weights later by removing the prior contribution since this is our importance weight; that is, our proposal for $s$ is $p(s)$ and we're targeting $p(s, m, y_{1:T})$ so our IS weight is
$$w = \frac{p(s, m, y_{1:T})}{p(s)} = \frac{p(s) p(m \mid s) p(y_{1:T} \mid s, m)}{p(s)} = p(m \mid s) p(y_{1:T} \mid s, m)$$

Does that make sense @yebai ?

And finally, whether we're resampling when hitting $\theta$ or not, is not a correctness issue since we're still targeting the same distribution (though it probably does something to the variance of the estimator; not sure if it's good or bad :shrug: )

Reply:
This also explains why we need to make `acclogp!!` accumulate log-probs in the task-local `varinfo` and not the "global" one.